Fix some failures in distributed mesh testing of porous flow

GiudGiud opened the pull request:
refs #25213

There are still 9 fails w/ distributed mesh (-p2 --distributed-mesh) after this PR so we cant turn on distributed mesh testing on porous flow yet. 

@cpgr

GiudGiud committed:
Fix some failures in distributed mesh testing of porous flow
refs #25213

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2758847/), step Docs: sync website on e5fcfda wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30140/site)

This comment will be updated on new commits.

cpgr commented:
Is the main problem that the id of the elements changes with `--distributed-mesh` so diffs when comparing the csv files?

GiudGiud commented:
There's that, there are objects that don't support distributed meshes, and there are differences in preconditioning that make the results change a bit. 

Seems there are a few tests that need attention in this PR still

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2761883/), step Generate coverage on 4784882 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30140/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented:
ok should be good now. There's about 10 failures left with -p 2 --distributed-mesh. 
If someone can figure them out (probably 2 or 3 common mode of failures) then we could turn on distributed mesh testing.

cpgr commented:
> ok should be good now. There's about 10 failures left with -p 2 --distributed-mesh. If someone can figure them out (probably 2 or 3 common mode of failures) then we could turn on distributed mesh testing.

Do I run them with `run_tests -p 2 --distributed-mesh`? And just in the porous_flow module? I'll see if I can fix them

GiudGiud commented:
> Do I run them with run_tests -p 2 --distributed-mesh? And just in the porous_flow module? 

yes to both

cpgr commented:
> > Do I run them with run_tests -p 2 --distributed-mesh? And just in the porous_flow module?
> 
> yes to both

OK. I'll have a go at them

cpgr committed:
Fix remaining distributed mesh test failures in porous flow

Most have been marked as replicated only due to objects not
compatible with distributed mesh testing

Refs #25213

cpgr commented:
Hi @GiudGiud 

I've worked through the remaining distributed mesh testing failures in porous flow.

One was fixed by tightening the relative tolerance.

I skipped (set mesh_mode = replicated) the failing tests in the heterogeneous_materials folder because they rely on a `RandomIC` to set an AuxVariable value. When run with `--distributed-mesh`, different random values are assigned, so there is no way to make these pass the distributed mesh testing as the output depends on these values.

I also skipped the remaining failures as it appears the combination of adaptivity, `NodalValueSampler` and `--distributed-mesh` results in different node ids, even when `allow_numbering = false` is set in the `Mesh` block.

My changes are at https://github.com/cpgr/moose/tree/PR-modules-testing. Don't know if you want to add them to this PR?

GiudGiud commented:
That's great news! 
I added the commit to this PR and we can use the sweep-porous flow recipe at p2 to check if it all went as expected.
Note that the sweep will test p2 p3 ... p16 so it might actually fail still, but distributed mesh -p2 is most of the work

cpgr commented:
Let me know if there are any other failures and I'll try and fix them

loganharbour commented:
> It's green at -p2 for regular and distributed meshes, but not for recover
> 
> @loganharbour can we turn on just those 2 parts and leave recover on the side for now?

@GiudGiud in the recipes that you enable, also set:
```
ONLY_ON_PR_NUM = 30140
```

then we can invalidate and test just here

GiudGiud commented:
it passed @loganharbour 
https://civet.inl.gov/job/2781164/

cpgr commented:
I've nearly fixed the failing recover tests now. Just one I haven't figured out

cpgr committed:
Fix failing recover tests in porous flow

Minor changes to the way these are tested rather than changes to
the test files themselves.

Refs #25213

cpgr commented:
I pushed another commit that fixes the failing recover tests except for one.

The test `test\tests\pressure_pulse\pressure_pulse_1d_fv.i` fails when using 

```
run_tests --recover
```

 but passes when doing 

```
porous_flow-opt -i pressure_pulse_1d_fv.i --test-checkpoint-half-transient
porous_flow-opt -i pressure_pulse_1d_fv.i --recover
```

Still can't figure out why

GiudGiud commented:
my guess is it's the test suite
```
  [pressure_pulse_1d_fv]
    type = 'CSVDiff'
    input = 'pressure_pulse_1d_fv.i'
    csvdiff = 'pressure_pulse_1d_fv.csv'
    rel_err = 1.0E-5
    threading = '!pthreads'
    issues = '#21275'
    design = 'porous_flow/tests/pressure_pulse/pressure_pulse_tests.md'
    requirement = 'The system shall correctly simulate the transient evolution of a pressure pulse in 1D using 1 phase physics with FV variables.'
  []
  [pressure_pulse_1d_fullys_saturated_fv]
    type = 'CSVDiff'
    input = 'pressure_pulse_1d_fully_saturated_fv.i'
    csvdiff = 'pressure_pulse_1d_fv.csv'
    rel_err = 1.0E-5
    threading = '!pthreads'
    issues = '#21275'
    design = 'porous_flow/tests/pressure_pulse/pressure_pulse_tests.md'
    requirement = 'The system shall correctly simulate the transient evolution of a pressure pulse in 1D using 1 phase fully-saturated physics with FV variables.'
  []
```

you have 2 files writing to the same CSV file

cpgr commented:
Damn. That was it. Adding a `prereq` in the test files fixes it. Pushed new change and all the recover tests work on my local machine

GiudGiud commented:
Awesome. I ll add the testing and see if it passes

moosebuild commented:
Job [Modules recover](https://civet.inl.gov/job/2787315/) on 4a660ad : invalidated by @GiudGiud

turned on porous flow recover testing

loganharbour reviewed (approved):
A +1 from @cpgr on the first commit would be good

GiudGiud commented:
no hsould not be
[1.943s]     DIFF time_steppers/time_adaptive.test FAILED (CSVDIFF)

we did not touch this

cpgr commented:
> A +1 from @cpgr on the first commit would be good

+1 on @GiudGiud commits!
